scipy/optimize/tests/test_least_squares.py

```diff
@@ -0,0 +1,466 @@
+            for tr_solver, jac in product(['exact', 'lsmr'],
+                                          ['2-point', '3-point']):
+                res_dense = least_squares(
+                    p.fun, p.x0, jac, method=self.method, tr_solver=tr_solver)
+                res_sparse = least_squares(
+                    p.fun, p.x0, jac, method=self.method, tr_solver=tr_solver,
+                    jac_sparsity=p.sparsity)
+                assert_equal(res_dense.nfev, res_sparse.nfev)
+                assert_allclose(res_dense.x, res_sparse.x, atol=1e-20)
+                assert_allclose(res_dense.cost, 0, atol=1e-20)
+                assert_allclose(res_sparse.cost, 0, atol=1e-20)
+
+    def test_with_bounds(self):
+        p = BroydenTridiagonal()
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", UserWarning)
+            for jac, tr_solver, jac_sparsity in product(
+                    [p.jac, '2-point', '3-point'],
+                    ['exact', 'lsmr'],
+                    [None, p.sparsity]):
+                res_1 = least_squares(p.fun, p.x0, jac, bounds=(p.lb, np.inf),
+                                      tr_solver=tr_solver, method=self.method,
+                                      jac_sparsity=jac_sparsity)
+                res_2 = least_squares(p.fun, p.x0, jac, bounds=(-np.inf, p.ub),
+                                      method=self.method, tr_solver=tr_solver,
+                                      jac_sparsity=jac_sparsity)
+                res_3 = least_squares(p.fun, p.x0, jac, bounds=(p.lb, p.ub),
+                                      method=self.method, tr_solver=tr_solver,
+                                      jac_sparsity=jac_sparsity)
+                assert_allclose(res_1.optimality, 0, atol=1e-10)
+                assert_allclose(res_2.optimality, 0, atol=1e-10)
+                assert_allclose(res_3.optimality, 0, atol=1e-10)
+
+    def test_wrong_jac_sparsity(self):
+        p = BroydenTridiagonal()
+        sparsity = p.sparsity[:-1]
+        assert_raises(ValueError, least_squares, p.fun, p.x0,
+                      jac_sparsity=sparsity, method=self.method)
+
+    def test_linear_operator(self):
+        p = BroydenTridiagonal(mode='linear_op')
+        res = least_squares(p.fun, p.x0, p.jac, method=self.method)
+        assert_allclose(res.cost, 0.0, atol=1e-20)
+        assert_raises(ValueError, least_squares, p.fun, p.x0, p.jac,
+                      method=self.method, tr_solver='exact')
+
+    def test_scaling(self):
+        p = BroydenTridiagonal()
+        res = least_squares(p.fun, p.x0, p.jac, method=self.method,
+                            scaling='jac')
+        assert_allclose(res.cost, 0.0, atol=1e-20)
+
+        p = BroydenTridiagonal(mode='linear_op')
+        assert_raises(ValueError, least_squares, p.fun, p.x0, p.jac,
+                      method=self.method, scaling='jac')
+
+
+class TestDogbox(TestCase, BaseMixin, BoundsMixin, SparseMixin):
```

Comment:
Python's MRO makes it natural to place mixins *before* the base class. Yes, it takes a bit to get used to :-).